### PR TITLE
Add check if the gift wrapping is disabled

### LINKED ISSUE #22825: Gift-wrapping line is displayed in order confirmation emails even if the gift option is disabled
### Describe the bug

I disabled the gift option from our shop settings because we don't provide this service, but emails still displays the `Gift-wrapping` line in the summary which is confusing.

### The expected behavior
Emails template should be updated to hide this line if the option is disabled (or if the value of this line is 0):
https://github.com/PrestaShop/PrestaShop/search?l=Twig&q=Gift-wrapping

### Steps to reproduce

1. **BO** > **CONFIGURE** > **Shop Parameters** > **Order Settings**
2. Scroll down to **Gift options** block
3. Disable the **Offer gift wrapping** option 
4. Save
5. FO > create an order
6. Open the **Order Confirmation** mail > See error

image

### Specs
When the option Offer gift wrapping is disabled on the Order Settings > Order settings page, the Gift-wrapping line in order_conf template is hidden.

### Additional infos
- PS version: 1.7.8.5
- PHP version: 7.2

### PR DESCRIPTION
**Pull request by marsaldev**
Don't show the line in the order summary. I've added a reminder hint in the BO form

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | If you disable the Gift Option, in the order summary the gift wrapping line is still present. It will be confusing.
| Type?             | bug fix
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #22825.
| How to test?      | Apply edits, regenerate e-mails.
| Possible impacts? | If the merchant enable or disable _Gift Wrapping_ option he/she needs to re-generate e-mails template. In this PR I've added a small hint in the BO as help text near the option

If this PR will be approved, I will made a PR in https://github.com/PrestaShop/mjml-theme-converter

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27890)

### CODE REVIEW

File: mails/themes/modern/core/order_conf.html.twig
Comment (matks) [inline mails/themes/modern/core/order_conf.html.twig:182]: I think the indentation is now broken. Can we fix it?

File: src/Adapter/MailTemplate/MailTemplateTwigRenderer.php
Comment (matks) [inline src/Adapter/MailTemplate/MailTemplateTwigRenderer.php]: This is a backward compatibility break because if someone is currently using this constructor, after your PR is merged it will now crash (fatal error) because the new constructor argument will be missing in his code.

But good news: next version is a major version! BC breaks are allowed 😄  . However we need to list them in the PR description.

Comment (matks) [reply src/Adapter/MailTemplate/MailTemplateTwigRenderer.php]: By the way PHPStan (see below) reported issues _exactly_ because of this problem: in some unit tests, MailTemplateTwigRenderer is being constructed. You did not update the code of these tests so these tests now crash.

**Review by matks — CHANGES_REQUESTED** [review]
Nice feature @marsaldev !

I reported a few issues, quite easy to fix. Also PHP CS Fixer is complaining (ref https://github.com/PrestaShop/PrestaShop/runs/5510708553?check_suite_focus=true). You can make him happy by running `php vendor/bin/php-cs-fixer fix`

File: mails/themes/modern/core/order_conf.html.twig
Comment (marsaldev) [reply mails/themes/modern/core/order_conf.html.twig:182]: which is the correct one? xD
I think the proposed one, although this file is generated by mjml 🤔 wdyt?

Comment (matks) [reply mails/themes/modern/core/order_conf.html.twig:182]: Mmmm ... interesting.
Do you think the fact there is blank space or not has an impact on how Mail clients would render the mail?

File: src/Adapter/MailTemplate/MailTemplateTwigRenderer.php
Comment (marsaldev) [reply src/Adapter/MailTemplate/MailTemplateTwigRenderer.php]: Thanks @matks! 
One question, I need to write the unit tests too?

Comment (matks) [reply src/Adapter/MailTemplate/MailTemplateTwigRenderer.php]: My opinion is that there is not a lot of value to be tested here. You give `$configuration` to the class, then it uses `get()` to retrieve `giftWrapping` and then we expose it as a smarty variable. We do not transform the data, we just transfer it.

File: mails/themes/modern/core/order_conf.html.twig
Comment (marsaldev) [reply mails/themes/modern/core/order_conf.html.twig:182]: 🤔 I don't think so, now we follow the table indentation, in theory there should be no issues on rendering the html generated

File: src/Adapter/MailTemplate/MailTemplateTwigRenderer.php
Comment (marsaldev) [reply src/Adapter/MailTemplate/MailTemplateTwigRenderer.php]: Got it, but I need to update the code in the existing unit tests, right?

Comment (matks) [reply src/Adapter/MailTemplate/MailTemplateTwigRenderer.php]: Yes. Existing unit tests need to be updated so that they pass

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php]:
```suggestion
                'help' => $this->trans(Remember to regenerate email templates after enabling or disabling this feature.', 'Admin.Shopparameters.Help'),
```

I removed the first sentence as it was a repetition of the label. Also, maybe we could specify which templates should be regenerated?

Comment (marsaldev) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php]: As far as I remember you can't chose a specific one, the regeneration process will regenerate all the e-mails (Design > Email theme)
Maybe we could add the path of the BO, like:
`Remember to regenerate email templates after enabling or disabling this option in Design > Email theme`

Comment (Julievrz) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php]: Good idea, then why not add a link? :-)

`Remember to regenerate email templates after enabling or disabling this option in [1]Design > Email theme[\1].`

Comment (marsaldev) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php]: Done 👍🏻

File: src/PrestaShopBundle/Resources/config/services/adapter/mail_template.yml
Comment (Progi1984) [inline src/PrestaShopBundle/Resources/config/services/adapter/mail_template.yml]:
```suggestion
      - "@=service('prestashop.adapter.legacy.configuration').getBoolean('PS_GIFT_WRAPPING')"
```

File: src/Adapter/MailTemplate/MailTemplateTwigRenderer.php
Comment (Progi1984) [inline src/Adapter/MailTemplate/MailTemplateTwigRenderer.php]:
```suggestion
    /** @var bool */
    private $hasGiftWrapping;
```

File: src/Adapter/MailTemplate/MailTemplateTwigRenderer.php
Comment (Progi1984) [inline src/Adapter/MailTemplate/MailTemplateTwigRenderer.php]:
```suggestion
        bool $hasGiftWrapping
```

File: src/Adapter/MailTemplate/MailTemplateTwigRenderer.php
Comment (Progi1984) [inline src/Adapter/MailTemplate/MailTemplateTwigRenderer.php]:
```suggestion
        $this->hasGiftWrapping = $hasGiftWrapping;
```

File: src/Adapter/MailTemplate/MailTemplateTwigRenderer.php
Comment (Progi1984) [inline src/Adapter/MailTemplate/MailTemplateTwigRenderer.php]:
```suggestion
        $layoutVariables['giftWrapping'] = $this->hasGiftWrapping;
```

**Review by Progi1984 — CHANGES_REQUESTED** [review]
@marsaldev Thank you for your contribution. Could you document the BC Break, please ?

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php]:
```suggestion
                'help' => $this->trans('Remember to regenerate email templates in [1]Design > Email theme[/1] after enabling or disabling this feature.',
```

Otherwise, it looks like the feature can be enabled/disabled in Design>Email theme

**Comment by marsaldev** [comment]
> @marsaldev Thank you for your contribution. Could you document the BC Break, please ?

Yes no problem, but where 😊 ?

~ ~ ~

I've accepted the requested changes, but I don't understand the real difference to provide the single value from the yml file rather than the solution proposed, give the Configuration to the constructor and then retrieve the value. 

I think it could be more useful to have the Configuration element available at class level also in the future, because if in the future we want add another setting we have to update the yml, the constructor and the tests as well. This will introduce another BC Break. It makes sense?
I don't think so 😊 

ping @Progi1984

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php
Comment (Progi1984) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php]:
```suggestion
     * @param string $defaultCurrencyIsoCode
```

File: tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php
Comment (Progi1984) [inline tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php]:
```suggestion
```

File: tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php
Comment (Progi1984) [inline tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php]:
```suggestion
        $generator = new MailTemplateTwigRenderer($engineMock, $builderMock, $dispatcherMock, false);
```

File: tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php
Comment (Progi1984) [inline tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php]:
```suggestion
```

File: tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php
Comment (Progi1984) [inline tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php]:
```suggestion
            false
```

File: tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php
Comment (Progi1984) [inline tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php]:
```suggestion
```

File: tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php
Comment (Progi1984) [inline tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php]:
```suggestion
            true
```

File: tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php
Comment (Progi1984) [inline tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php]:
```suggestion
            true
```

File: tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php
Comment (Progi1984) [inline tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php]:
```suggestion
            true
```

File: tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php
Comment (Progi1984) [inline tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php]:
```suggestion
            true
```

File: tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php
Comment (Progi1984) [inline tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php]:
```suggestion
            true
```

**Review by Progi1984 — DISMISSED** [review]
@marsaldev Thank you for your contribution. I have brought you some comments about your test for fixing them.
Could you document the BC Break and add a linked PR on the devdocs : https://devdocs.prestashop.com/8/modules/core-updates/8.0/ ?

**Comment by marsaldev** [comment]
> @marsaldev Thank you for your contribution. I have brought you some comments about your test for fixing them. Could you document the BC Break and add a linked PR on the devdocs : https://devdocs.prestashop.com/8/modules/core-updates/8.0/ ?

Created a PR for the devdocs 😉 
Applied your suggestions + applied phpcsfixer suggestion

**Review by atomiix — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @marsaldev 

I have checked this PR. The issue is not fixed.

1. Disable the option `Offer gift wrapping` from BO > Shop Parameters > Order settings > Gift options
2. Regenerate email template from Design > Email Theme (With Modern and classic theme)
3. Create an order from BO 
4. Check the order summary of the **new order** received in the email
5. The line `Gift-wrapping` still exist 

Here is the example : 
image
image
image
image

https://user-images.githubusercontent.com/92912932/162754317-37f75a3c-3ab5-4f85-a73b-195378b98de4.mp4


The same for modern or classic theme 
image

https://user-images.githubusercontent.com/92912932/162753493-66f7720a-a3a6-407b-b0df-ebf04e41e62f.mp4

Could you please check it ?

Thanks!

**Comment by Seb33300** [comment]
Hello @marsaldev, maybe you can also add the same changes for the `recycled packaging` option which has the exact same issue.

I think this is a new option introduced in PrestaShop 1.7.8
It works in the exact same way as the gift wrapping option.

https://github.com/PrestaShop/PrestaShop/blob/7f058bc7e7a9a5f37959b4d7b492b64daa42a2b9/mails/themes/modern/core/order_conf.html.twig#L309

**Comment by Seb33300** [comment]
@HanaRebaiQA you need to check `Overwrite templates` option when regenerating emails templates.
Otherwise the email template will not be updated.

image

**Comment by HanaRebaiQA** [comment]
Hello @Seb33300 Yes! the `Overwrite templates` is well enabled (please check my firt screen-record)
I specify that the issue appears in the **new_order** email 

Thanks!

**Comment by Seb33300** [comment]
Can you check that you do not have email template generated in your theme folder?
Theme folder has the priority on the core folder.

If your theme has email template folder, you should select your theme in the list instead of core.

**Comment by marsaldev** [comment]
> Hello @marsaldev, maybe you can also add the same changes for the `recycled packaging` option which has the exact same issue.
> 
Hi @Seb33300 
Sure, is there already an issue opened?

Hi @HanaRebaiQA 
I'll make a check asap 😉 but this PR is about the _order_conf_ template, the e-mail read by the customer and not by the merchant 😊  so I'd say that is normal to still have the gift wrapping line in the _new_order_ e-mail 😊

**Comment by Seb33300** [comment]
> Hi @Seb33300
Sure, is there already an issue opened?

I don't think there is a specific issue for the recycled package option.
I did not put it in [my issue](https://github.com/PrestaShop/PrestaShop/issues/22825) because recycled package option did not exist when I created it.

Now that I have upgraded my project to 1.7.8, I am facing it and for me, the issue I described for gift wrapping is the same for recycled package.

**Comment by marsaldev** [comment]
Hi @HanaRebaiQA, 
just checked, is like you showed, but as I said, this PR is for `order_conf` e-mail, and not the e-mail received to merchant, sent from `ps_emailalerts` module 😊

File: src/Adapter/MailTemplate/MailTemplateTwigRenderer.php
Comment (atomiix) [inline src/Adapter/MailTemplate/MailTemplateTwigRenderer.php:75]: Can you update the phpdoc?

Comment (marsaldev) [reply src/Adapter/MailTemplate/MailTemplateTwigRenderer.php:75]: Hi @atomiix,
if you're referring to the BC I've already opened a PR: https://github.com/PrestaShop/docs/pull/1333 😊

Comment (atomiix) [reply src/Adapter/MailTemplate/MailTemplateTwigRenderer.php:75]: No sorry, I was talking about lines 64-66 😅:
```php
/**
 * @param Environment $twig
 * @param LayoutVariablesBuilderInterface $variablesBuilder
 * @param HookDispatcherInterface $hookDispatcher
```

Comment (marsaldev) [reply src/Adapter/MailTemplate/MailTemplateTwigRenderer.php:75]: Done 😉

**Review by Progi1984 — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @marsaldev 

Thanks for your feedback!

I have checked again this PR. I still have the issue with `order_conf `when use the `classic` email theme.

Here the steps to reproduce : 

1. BO > CONFIGURE > Shop Parameters > Order Settings
2. Scroll down to Gift options block
3. Disable the `Offer gift wrapping` option and save
4. BO > IMPROVE > Design > Email Theme > Configuration > Select `classic` as the default email theme
5. In Generate emails > Select `classic` , Select the default language , Enable the `Overwrite templates `
6. Click on generate emails
7. In Email themes > Click to Preview the classic theme
8. Search for the `order_conf` and Preview the template with Http 
9. See error > The ` Gift-wrapping`  still displayed
10. Go to FO > Create an order
11. Open the Order Confirmation mail 
12. See error > The ` Gift-wrapping` still displayed

When choose the `modern` email theme, the` Gift-wrapping` is not displayed in the email confirmation 

image

https://user-images.githubusercontent.com/92912932/164027483-be81f287-a30b-4c6e-9ac8-616628912b41.mp4



Could you please check it ?

Thanks!

**Comment by marsaldev** [comment]
Hi @HanaRebaiQA,
thanks a lot for your testing! 
I think it's "normal" because the edits are only on _modern_ template ^^"
I'm adding the fix also for the _classic_ with a new commit 😉

**Review by atomiix — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @marsaldev 

I have checked again this PR and the issue with `classic` theme that i mentioned [here](https://github.com/PrestaShop/PrestaShop/pull/27890#pullrequestreview-945816281) is not fixed.

image


https://user-images.githubusercontent.com/92912932/165527932-5a900e6e-ffd7-4dac-828f-1e98b6f42182.mp4


Could you please check again ?

Thanks!

**Comment by marsaldev** [comment]
Hi @HanaRebaiQA, thanks for your testing!
Sorry, but I'm not able to reproduce your behaviour, see my recording where I enabled and disabled two times the gift wrapping 😊 
(without errors :()

https://user-images.githubusercontent.com/48283112/165726271-42c20d55-e864-41ab-9c89-948254ae07da.mp4

**Comment by kpodemski** [comment]
Hello @marsaldev @HanaRebaiQA 

Maybe pair-coding, pair-testing session would be a good idea to proceed with this PR 👍

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @marsaldev 

Thanks for your feedback. I also have checked again and i still have the issue with `classic`  email theme.

Thanks @kpodemski and Im OK for check with @marsaldev 

https://user-images.githubusercontent.com/92912932/170501149-25722827-c89b-4e69-8f03-a9876a8b47e1.mp4


Thank you!

File: mails/themes/classic/core/order_conf.html.twig
Comment (Seb33300) [inline mails/themes/classic/core/order_conf.html.twig]: You put the condition on the shipping line.

**Comment by Seb33300** [comment]
@marsaldev you made a mistake on the classic template. The shipping cost line is hidden instead of the gift wrapping line.

(if you look at the video you can see that the Shipping line in totals is hidden instead of the gift wrapping line)

https://github.com/PrestaShop/PrestaShop/blob/83dddca1715f90b03bf8e3501fbf440c82917a6f/mails/themes/classic/core/order_conf.html.twig#L149-L178

**Review by Seb33300 — CHANGES_REQUESTED** [review]

File: mails/themes/classic/core/order_conf.html.twig
Comment (marsaldev) [reply mails/themes/classic/core/order_conf.html.twig]: Oh my, you're right!

**Review by atomiix — APPROVED** [review]

**Review by kpodemski — APPROVED** [review]

**Review by HanaRebaiQA — APPROVED** [review]
Hello @marsaldev 

Thanks for the fixed issue. I have checked again this PR and here is my tested cases : 

-Automated tests : OK 
-Check when `Gift-wrapping` is enabled AND when it's disabled with all combinations of `classic` and `modern` email-theme : OK
-Check when create an order from BO : OK 
-Check when create an order from FO : OK 
-Check email-theme generated from Http from the preview theme list of `modern` and `classic` : OK  
-Check with multistore : Ok 
-Check with different langauges : OK
-Check with `maildev` server : OK  

**Modern email-theme when Gift-wrapping is disabled** :heavy_check_mark: 
image

**Classic email-theme when Gift-wrapping is disabled** :heavy_check_mark: 
image


So, it's QA :heavy_check_mark: 

Thank you!

**Review by Seb33300 — APPROVED** [review]

**Comment by Progi1984** [comment]
Thanks @marsaldev & @HanaRebaiQA

**Comment by Progi1984** [comment]
@PrestaShop/prestashop-maintainers Is this PR needs to be reverted as it brings "BC Breaks"?

**Comment by Progi1984** [comment]
@PrestaShop/prestashop-core-developers  Is this PR needs to be reverted as it brings "BC Breaks"?

**Comment by kpodemski** [comment]
@Progi1984 let's discuss this on Tuesday during a meeting

**Comment by Progi1984** [comment]
@kpodemski Please tell me the decision you took after this one.

**Pull request merged**
